platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/DefaultIntakeEventCounter.java

```
@@ -0,0 +1,105 @@
+     * but prevents it from going below 0.
+     */
+    private static final IntUnaryOperator EXIT_INTAKE = count -> count > 0 ? count - 1 : 0;
+
+    /**
+     * A map from node id to an atomic integer, which represents the number of events from that peer that have been
+     * added to the intake pipeline, but haven't yet made it through.
+     */
+    private final Map<NodeId, AtomicInteger> unprocessedEventCounts;
+
+    /**
+     * Constructor
+     *
+     * @param addressBook the address book
+     */
+    public DefaultIntakeEventCounter(@NonNull final AddressBook addressBook) {
+        this.unprocessedEventCounts = new HashMap<>();
+
+        for (final NodeId nodeId : addressBook.getNodeIdSet()) {
+            unprocessedEventCounts.put(nodeId, new AtomicInteger(0));
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean hasUnprocessedEvents(@NonNull final NodeId peer) {
+        Objects.requireNonNull(peer);
+
+        return unprocessedEventCounts.get(peer).get() > 0;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void eventEnteredIntakePipeline(@NonNull final NodeId peer) {
+        Objects.requireNonNull(peer);
+
+        unprocessedEventCounts.get(peer).incrementAndGet();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void eventExitedIntakePipeline(@Nullable final NodeId peer) {
+        if (peer == null) {
+            // this will happen if the event wasn't received through normal gossip
+            return;
+        }
+
+        if (unprocessedEventCounts.get(peer).getAndUpdate(EXIT_INTAKE) == 0) {
+            throw new IllegalStateException(
+                    "Event processed from node %s, but no known events from that peer were in the intake pipeline. This shouldn't be possible."
+                            .formatted(peer));
+        }
```

I'd suggest moving the error handling logic into the `EXIT_INTAKE` function. I think that would make it easier to reason about when the exception will be thrown. 

```
private static final IntUnaryOperator EXIT_INTAKE = count -> count > 0 ? count - 1 : throw new IllegalStateException(...);
```

---

Look at `AbstractReservable.java`, I throw exceptions in `IntUnaryOperator` lambdas there. We can't throw a checked exception (i.e. one you are forced to catch and have to put in the function interface), but we can throw an unchecked exception. `IllegalStateException` is unchecked.

---

Done